### PR TITLE
Partial Fix #1412: Enhance java.io.File to support URI

### LINKED ISSUE #1412: Unable to link sconfig due to missing Java classes
I am currently working on getting the `lightbend/config` port, now `ekrich/sconfig` to compile under Scala Native and here is my current output. The goal is to support `scalafmt` on Scala Native.

I will look to see if any of these are internal details. The `java.time.Duration` class is part of the interface.

cc @olafurpg

List of items in the works for a possible `0.3.9` release. I will provide PRs once a branch is added.

- [x] Vector and Stack fixed by using `scala.collection.mutable.ArrayStack` in sconfig
- [x] Forwarders and `trait` additions
- [x] Adding `java.util.IdentityHashMap`
- [x] Adding tests for `IdentityHashMap`
- [x] Adding methods and ctor to `java.io.File`
- [x] Adding tests for `File`
- [x] Adding minimum `java.time` classes and methods
- [x] Adding tests for `java.time` additions
- [x] URL and URLConnection stubs

```
[IJ]sbt:sconfig-test> run
[info] Compiling 2 Scala sources to /Users/eric/workspace/sconfig-test/target/scala-2.11/classes ...
[info] Done compiling.
[info] Linking (2709 ms)
[error] cannot link: @java.io.File::init_java.net.URI
[error] cannot link: @java.io.File::toURI_java.net.URI
[error] cannot link: @java.lang.Class$::forName_java.lang.String_java.lang.Class
[error] cannot link: @java.lang.Class::getConstructor_scala.scalanative.runtime.ObjectArray_java.lang.reflect.Constructor
[error] cannot link: @java.lang.Class::getField_java.lang.String_java.lang.reflect.Field
[error] cannot link: @java.lang.ClassLoader::getResources_java.lang.String_java.util.Enumeration
[error] cannot link: @java.lang.ClassLoader::loadClass_java.lang.String_java.lang.Class
[error] cannot link: @java.lang.Thread::getContextClassLoader_java.lang.ClassLoader
[error] cannot link: @java.lang.reflect.Constructor::newInstance_scala.scalanative.runtime.ObjectArray_java.lang.Object
[error] cannot link: @java.net.URL::toExternalForm_java.lang.String
[error] cannot link: @java.net.URL::toURI_java.net.URI
[error] cannot link: @java.net.URLConnection::connect_unit
[error] cannot link: @java.net.URLConnection::getContentType_java.lang.String
[error] cannot link: @java.net.URLConnection::getInputStream_java.io.InputStream
[error] cannot link: @java.net.URLConnection::setRequestProperty_java.lang.String_java.lang.String_unit
[error] cannot link: @java.time.Duration
[error] cannot link: @java.time.Duration::toMillis_i64
[error] cannot link: @java.util.ArrayList::listIterator_i32_java.util.ListIterator
[error] cannot link: @java.util.IdentityHashMap
[error] cannot link: @java.util.IdentityHashMap::init
[error] cannot link: @java.util.List::containsAll_java.util.Collection_bool
[error] cannot link: @java.util.List::equals_java.lang.Object_bool
[error] cannot link: @java.util.List::hashCode_i32
[error] cannot link: @java.util.List::removeAll_java.util.Collection_bool
[error] cannot link: @java.util.List::remove_java.lang.Object_bool
[error] cannot link: @java.util.Properties::load_java.io.Reader_unit
[error] cannot link: @java.util.Stack
[error] cannot link: @java.util.Stack::init
[error] cannot link: @java.util.Stack::pop_java.lang.Object
[error] cannot link: @java.util.Stack::push_java.lang.Object_java.lang.Object
[error] cannot link: @java.util.Vector
[error] cannot link: @java.util.Vector::isEmpty_bool
[error] unable to link
```

### PR DESCRIPTION
**Pull request by ekrich**
This is the 3rd PR of 5 to support [sconfig](https://github.com/ekrich/sconfig) and then native `scalafmt`.
See https://github.com/scala-native/scala-native/issues/1412

This PR is targeted for 0.4.0 and 0.3.9.

### CODE REVIEW

**Comment by densh** [comment]
@ekrich is this the last needed for scalafmt to link?

**Comment by ekrich** [comment]
@densh I have a Duration PR ready but not linking for test yet and I need some URL stubs to make it link.

**Comment by ekrich** [comment]
@densh Ready for review.

File: javalib/src/main/scala/java/io/File.scala [hunk 1]
```diff
@@ -541,4 +571,22 @@ object File {
                                minLength = true,
                                throwOnError = true)
 
+  // Ported from Apache Harmony
+  private def checkURI(uri: URI): Unit = {
+    def throwExc(msg: String): Unit =
+      throw new IllegalArgumentException(s"$msg: $uri")
+    def compMsg(comp: String): String = s"Found $comp component in URI"
```
Comment (densh) [inline javalib/src/main/scala/java/io/File.scala]: Please put a newline after the `=` here.

File: javalib/src/main/scala/java/io/File.scala [hunk 2]
```diff
@@ -541,4 +571,22 @@ object File {
...
+    else if (!uri.getRawSchemeSpecificPart.startsWith("/"))
+      throwExc("URI is not hierarchical")
+    else if (uri.getScheme == null || !(uri.getScheme == "file"))
+      throwExc("Expected file scheme in URI")
+    else if (uri.getRawPath == null || uri.getRawPath.length == 0)
+      throwExc("Expected non-empty path in URI")
+    else if (uri.getRawAuthority != null) throwExc(compMsg("authority"))
+    else if (uri.getRawQuery != null) throwExc(compMsg("query"))
+    else if (uri.getRawFragment != null) throwExc(compMsg("fragment"))
+    // else URI is ok
```
Comment (densh) [inline javalib/src/main/scala/java/io/File.scala]: Please wrap if/else branch bodies into `{ }` to improve readability.

File: javalib/src/main/scala/java/io/File.scala [hunk 3]
```diff
@@ -323,8 +327,34 @@ class File(_path: String) extends Serializable with Comparable[File] {
...
+        new URI("file", "", path, null)
+      } else {
+        new URI("file", null, path, null, null)
+      }
+
+    } catch {
+      case e: URISyntaxException =>
+        // this should never happen
+        return null
+    }
```
Comment (densh) [inline javalib/src/main/scala/java/io/File.scala]: If this should never happen why is it in the code?

File: javalib/src/main/scala/java/io/File.scala (hunk 1 above)
Comment (ekrich) [reply javalib/src/main/scala/java/io/File.scala]: I also added a blank line before the body of the method.

File: javalib/src/main/scala/java/io/File.scala (hunk 2 above)
Comment (ekrich) [reply javalib/src/main/scala/java/io/File.scala]: Yes, and done.

File: javalib/src/main/scala/java/io/File.scala (hunk 3 above)
Comment (ekrich) [reply javalib/src/main/scala/java/io/File.scala]: Blind porting from Harmony - removed.

**Comment by densh** [comment]
LGTM, thanks @ekrich ! 👍

**Comment by densh** [comment]
LGTM, thanks @ekrich ! 👍

**Pull request merged**